README.md
# vandium-serverless

[![serverless](http://public.serverless.com/badges/v3.svg)](http://www.serverless.com)
[![Build Status](https://travis-ci.org/vandium-io/vandium-serverless.svg?branch=master)](https://travis-ci.org/vandium-io/vandium-serverless)
[![npm version](https://badge.fury.io/js/vandium-serverless.svg)](https://badge.fury.io/js/vandium-serverless)

[Serverless](https://www.serverless.com) plugin allowing you to create new Node.js 4.3 functions automatically wrapped with [vandium](http://vandium.io).

## Features

* Creates Serverless functions pre-wrapped with vanidum
* Adds HTTP response codes specific to vandium error conditions
* No need to remember wrap your Serverless functions in vandium manually
* Easy to use
* Integrates with Serverless Framework for AWS Lambda
* Supports Node.js 4.3.2

## Installation

* Install via npm in the root folder of your Serverless project.

```
npm install vandium-serverless --save
```

* Edit your ```s-project.json``` to include the ```vandium-serverless``` plugin.

```json
{
  "name": "yourprojectname",
  "custom": {},
  "plugins": [ "vandium-serverless" ]
}
```

## Getting Started

Run ```serverless function create``` to create a new function, as you would normally do with Serverless.

When prompted to select a runtime for the new function select 'nodejs4.3-vandium'.

```
Serverless: Please, select a runtime for this new Function
    nodejs4.3
    python2.7
  > nodejs4.3-vandium
    nodejs (v0.10, soon to be deprecated)
```

Your newly created function is a standard Node.js 4.3 function wrapped with vandium.

```js
'use strict';

const vandium = require( 'vandium' );

/*
vandium.validation( {

    // your validation code here
    // firstName: vandium.types.string().min( 4 ).max( 80 ).required(),
    // lastName: vandium.types.string().min( 4 ).max( 80 ).required(),
    // age: vandium.types.number().integer().min( 0 ).max( 120 )
});
*/

module.exports.handler = vandium( function( event, context, callback ) {

    // your code goes here

    callback( null, 'Your Vandium wrapped Serverless function ran succesfully!!!' );
});
```
Vandium offers features such as input validation, SQL Injection detection, and JWT authentication. For a full list of features and instructions visit the  [vandium](http://vandium.io) project page.

## Packaging

In order to use the Vandium wrapper, Vandium must be installed as a third party dependency.  The ```require( 'vandium' )``` is already included at the top of the template Lambda function code.  How you handle third party dependencies in your Serverless project is up to you, however here are several options based on the [Serverless documentation](http://docs.serverless.com/docs/function-configuration).

* Create a ```package.json``` file for each Lambda function, and run ```npm install vandium --save``` in each function directory.  This option results in smaller Lambda functions, but multiple ```node_modules``` and ```package.json``` paths to maintain.

* Use the ```package.json``` in the root directory of your Serverless project, and run ```npm install vandium --save``` there.  To use this option you will have to modify the ```handler``` property in the ```s-function.json``` files to include the full path to the function handler starting from the subdirectory in the project root directory.  It will look something like ```functionName/handler.handler``` or ```lib/functionName/handler.handler``` depending on how you store your functions.  This option will result in larger Lambda functions, since everything will be included in the Lambda function packaging, however there will only be one location for all dependencies (```node_modules``` and ```package.json```).

* You can store your dependencies anywhere you decide to place an additional ```node_modules``` and ```package.json```. Just remember your ```handler``` property must include a path starting in the directory where the dependencies are stored. That entire directory will get packaged when your function is deployed.  As an example,

```
parent
|
|-- functionName
|   |
|   |-- event.json
|   |
|   |-- handler.js
|   |
|   +-- s-function.json
|
|-- node_modules
|
+-- package.json
```

would require the handler property in ```s-function.json``` to look like ```"handler": "functionName/handler.handler"```

## Feedback

We'd love to get feedback on how you're using vandium-serverless and things we could add to make this tool better. Feel free to contact us at ```feedback@vandium.io```

## License

[BSD-3-Clause](https://en.wikipedia.org/wiki/BSD_licenses)


CHANGELOG.md
# Change Log

## 1.1.0 (2016-07-07)

Updated:
* Added HTTP response codes specific to vandium error conditions

Fixed:
* Changed output runtime in `s-function.json` from `nodejs4.3-vandium` to `nodejs4.3` for compatibility with `serverless-offline` package

## 1.0.0 (2016-05-16)

Initial Release


test/index.test.js
'use strict';
/*jshint expr: true*/


const sinon = require( 'sinon' );

const proxyquire = require( 'proxyquire' );

const expect = require( 'chai' ).expect;


describe( 'index.js', function() {

    let runtimeNode43VandiumStub;

    let indexModule;

    let addHookStub = sinon.stub();

    let readFileSyncStub = sinon.stub().returns( { runtime: 'nodejs4.3' } );

    let writeFileSyncStub = sinon.stub();

    let getRootPathStub = sinon.stub().returns( '/this/is/a/full/path' );

    const evt = { options: { path: '/this/is/a/path' } };

    let serverlessStub = {

        classes: {

            Plugin: class Plugin {}
        },

        addHook: addHookStub,

        utils: {

            readFileSync: readFileSyncStub,
            writeFileSync: writeFileSyncStub
        },

        getProject: function() {

            return {

                getRootPath: getRootPathStub
            }
        }
    };


    beforeEach( function() {

        runtimeNode43VandiumStub = sinon.stub().returns( 'runtimeNode43VandiumStub' );

        indexModule = proxyquire( '../index', {

            './lib/RuntimeNode43Vandium': runtimeNode43VandiumStub
        });
    });


    it( 'set the serverless classes to our new classes', function() {

        indexModule( serverlessStub );

        expect( serverlessStub.classes ).to.exist;
        expect( serverlessStub.classes.RuntimeNode43Vandium ).to.exist;
        expect( serverlessStub.classes.RuntimeNode43Vandium ).to.equal( 'runtimeNode43VandiumStub' );
    });


    it( 'create a new instance of our plugin', function() {

        let vandiumServerless =  new ( indexModule( serverlessStub ) ) ();

        expect( vandiumServerless.name ).to.equal( 'VandiumServerless' );
    });


    it( 'add our plugin as a hook after the function create action has run', function( done ) {

        let vandiumServerless = new ( indexModule( serverlessStub ) ) ();

        vandiumServerless.registerHooks();

        expect( addHookStub.calledOnce ).to.be.true;
        expect( addHookStub.firstCall.args[ 1 ] ).to.eql( { action: 'functionCreate', event: 'post' } );

        done();
    });


    it( 'run our plugin hook method without vandium', function() {

        let vandiumServerless = new ( indexModule( serverlessStub ) ) ();

        vandiumServerless._hookPost( evt );

        expect( getRootPathStub.calledOnce ).to.be.true;
        expect( getRootPathStub.firstCall.args[ 0 ] ).to.equal( evt.options.path );
        expect( getRootPathStub.firstCall.args[ 1 ] ).to.equal( 's-function.json' );
        expect( readFileSyncStub.calledOnce ).to.be.true;
        expect( readFileSyncStub.firstCall.args[ 0 ] ).to.equal( '/this/is/a/full/path' );

        expect( writeFileSyncStub.lastCall.args[ 1 ] ).to.eql( { runtime: 'nodejs4.3' } );
    });


    it( 'run our plugin hook method with vandium but no endpoint', function() {

        serverlessStub.utils.readFileSync = sinon.stub().returns( { runtime: 'nodejs4.3-vandium', endpoints: [] } );

        let vandiumServerless = new ( indexModule( serverlessStub ) ) ();

        vandiumServerless._hookPost( evt );

        expect( writeFileSyncStub.lastCall.args[ 1 ] ).to.eql( { runtime: 'nodejs4.3', endpoints: [] } );
    });


    it( 'run our plugin hook method with vandium and an endpoint', function() {

        serverlessStub.utils.readFileSync = sinon.stub().returns( {

            runtime: 'nodejs4.3-vandium',
            endpoints: [ {
                responses: {
                    '400': {
                        statusCode: '400'
                    }
                }
            }]
        });

        let http403 = {

            selectionPattern: '^authentication error.*',
            statusCode: '403'
        };

        let http422 = {

             selectionPattern: '^validation error.*',
             statusCode: '422'
        };

        let vandiumServerless = new ( indexModule( serverlessStub ) ) ();

        vandiumServerless._hookPost( evt );

        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ] ).to.exist;
        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ].responses ).to.exist;
        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ].responses[ '400' ] ).to.eql( { statusCode: '400' } );
        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ].responses[ '403' ] ).to.eql( http403 );
        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ].responses[ '422' ] ).to.eql( http422 );
    });

    it( 'run our plugin hook method with vandium and a empty endpoint', function() {

        serverlessStub.utils.readFileSync = sinon.stub().returns( {

            runtime: 'nodejs4.3-vandium',
            endpoints: [ {} ]
        });

        let http403 = {

            selectionPattern: '^authentication error.*',
            statusCode: '403'
        };

        let http422 = {

             selectionPattern: '^validation error.*',
             statusCode: '422'
        };

        let vandiumServerless = new ( indexModule( serverlessStub ) ) ();

        vandiumServerless._hookPost( evt );

        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ] ).to.exist;
        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ].responses ).to.exist;
        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ].responses[ '403' ] ).to.eql( http403 );
        expect( writeFileSyncStub.lastCall.args[ 1 ].endpoints[ 0 ].responses[ '422' ] ).to.eql( http422 );
    });
});


test/templates/handler.test.js
'use strict';
/*jshint expr: true*/

const proxyquire = require( 'proxyquire' );

const expect = require( 'chai' ).expect;

const sinon = require( 'sinon' );

const DEFAULT_MESSAGE = 'Your Vandium wrapped Serverless function ran succesfully!!!';

describe( 'hanlder.js', function() {


    let handlerModule;

    let vandiumStub;

    before( function() {

        vandiumStub = sinon.stub().returnsArg( 0 );

        handlerModule = proxyquire( '../../templates/handler.js', {

            'vandium': vandiumStub
        });
    });

    it( 'normal operation', function() {

        let event = {};

        let context = {};

        let callback = sinon.stub();

        handlerModule.handler( event, context, callback );

        expect( callback.calledOnce ).to.be.true;
        expect( callback.withArgs( null, DEFAULT_MESSAGE ).calledOnce ).to.be.true;
    });
});


test/lib/EndpointVandium.test.js
'use strict';
/*jshint expr: true*/

const expect = require( 'chai' ).expect;

const endpointVandium = require( '../../lib/EndpointVandium' );


class Endpoint {

    constructor() {

        this.responses = {};
    }
}


describe( 'EndpointVandium', function() {

    let serverless = {

        classes: {

            Endpoint
        }
    };

    it( 'nodejs4.3-vandium runtime', function() {

        let pluginClass = endpointVandium( serverless );

        expect( pluginClass.name ).to.equal( 'EndpointVandium' );

        let func = { runtime: 'nodejs4.3-vandium' };

        let plugin = new pluginClass( undefined, func );

        let http403 = {

            selectionPattern: '^authentication error.*',
            statusCode: '403'
        };

        let http422 = {

            selectionPattern: '^validation error.*',
            statusCode: '422'
        };

        expect( plugin ).to.be.an.instanceof( Endpoint );
        expect( plugin.responses ).to.exist;
        expect( plugin.responses[ '403' ]).to.exist;
        expect( plugin.responses[ '403' ]).to.eql( http403 );
        expect( plugin.responses[ '422' ]).to.exist;
        expect( plugin.responses[ '422' ]).to.eql( http422 );
    });

    it( 'other runtime', function() {

        let pluginClass = endpointVandium( serverless );

        expect( pluginClass.name ).to.equal( 'EndpointVandium' );

        let func = { runtime: 'nodejs4.3' };

        let plugin = new pluginClass( undefined, func );

        expect( plugin ).to.be.an.instanceof( Endpoint );
        expect( plugin.responses ).to.exist;
        expect( plugin.responses[ '403' ]).to.not.exist;
        expect( plugin.responses[ '422' ]).to.not.exist;
    });
});


lib/EndpointVandium.js
'use strict';

const AUTHENTICATION_FAILURE_CODE = '403';

const VALIDATION_FAILURE_CODE = '422';

module.exports = function( S ) {

    return class EndpointVandium extends S.classes.Endpoint {

        constructor( data, func ) {

            super( data, func );

            let _this = this;

            if( func.runtime === 'nodejs4.3-vandium' && !data ) {

                _this.responses[ AUTHENTICATION_FAILURE_CODE ] = {

                    selectionPattern: '^authentication error.*',
                    statusCode: AUTHENTICATION_FAILURE_CODE
                };

                _this.responses[ VALIDATION_FAILURE_CODE ] = {

                    selectionPattern: '^validation error.*',
                    statusCode: VALIDATION_FAILURE_CODE
                }
            }
        }
    }
}


lib/RuntimeNode43Vandium.js
'use strict';

const Promise = require( 'bluebird' );

const fs = Promise.promisifyAll( require( 'fs' ) );

const path = require( 'path' );

module.exports = function( S ) {

    return class RuntimeNode43Vandium extends S.classes.RuntimeNode43 {

        static getName() {

            return 'nodejs4.3-vandium';
        }


        getName( providerName ) {

            if( providerName === 'aws' ) {

                return 'nodejs4.3';
            }
            else {

                return RuntimeNode43Vandium.getName();
            }
        }

        /**
        * Scaffold
        * - Create scaffolding for new Node.js function
        */
        scaffold( func ) {

            const handlerPath = path.resolve( __dirname, '..', 'templates', 'handler.js' );

            func.handler = 'handler.handler';

            return fs.readFileAsync( handlerPath )
                .then( function( handlerJs ) {

                    return  Promise.all([

                        func.save(),
                        S.utils.writeFile( func.getRootPath( 'handler.js' ), handlerJs ),
                        S.utils.writeFile( func.getRootPath( 'event.json' ), {} )
                    ])
                });
        }

    }
}


index.js
'use strict';

const Promise = require( 'bluebird' );

const AUTHENTICATION_FAILURE_CODE = '403';

const VALIDATION_FAILURE_CODE = '422';

module.exports = function( S ) {

    S.classes.RuntimeNode43Vandium = require( './lib/RuntimeNode43Vandium' ) ( S ) ;

    return class VandiumServerless extends S.classes.Plugin {

        constructor() {

            super();
            this.name = 'VandiumServerless';
        }

        registerHooks() {

            S.addHook( this._hookPost.bind( this ), {

                action: 'functionCreate',
                event:  'post'
            });

            return Promise.resolve();
        }

        _hookPost( evt ) {

            return new Promise( function( resolve, reject ) {

                const filePath = S.getProject().getRootPath( evt.options.path, 's-function.json' );

                let savedFunction = S.utils.readFileSync( filePath );

                if( savedFunction.runtime === 'nodejs4.3-vandium' ) {

                    if( savedFunction.endpoints.length > 0 ) {

                        if( !savedFunction.endpoints[ 0 ].responses ) {

                            savedFunction.endpoints[ 0 ].responses = {};
                        }

                        savedFunction.endpoints[ 0 ].responses[ AUTHENTICATION_FAILURE_CODE ] = {

                            selectionPattern: '^authentication error.*',
                            statusCode: AUTHENTICATION_FAILURE_CODE
                        };

                        savedFunction.endpoints[ 0 ].responses[ VALIDATION_FAILURE_CODE ] = {

                            selectionPattern: '^validation error.*',
                            statusCode: VALIDATION_FAILURE_CODE
                        }
                    }

                    savedFunction.runtime = 'nodejs4.3';
                }

                S.utils.writeFileSync( filePath, savedFunction );

                return resolve( evt );

            });
        }
    }
}
